### PR TITLE
Fix for Clang + pager + ASLR (QEMUv8)

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

**Comment by jforissier** [comment]
This one was quite hard to troubleshoot actually :-/

I decided not to call it a linker bug, although it does look like one IMO. The [rules for absolute vs. relative symbols](https://sourceware.org/binutils/docs/ld/Expression-Section.html) are not that clear in my opinion... or maybe I did not spend enough time reading them. Anyways, once I had spotted the source of the issue it was only a matter of finding a syntax that works for both GCC and Clang. Doing "newsym = oldsym + value" was natural and more likely to retain the relocatable nature of the original symbol.

File: core/arch/arm/kernel/kern.ld.S [hunk 1]
```diff
@@ -61,6 +60,14 @@
 #define TEE_TEXT_VA_START       (TEE_RAM_VA_START + \
 					(TEE_LOAD_ADDR - TEE_RAM_START))
 
+/*
+ * Note:
+ * Clang 11 (ld.lld) generates non-relocatable references to symbols defined
+ * with ROUNDUP()/ROUNDDOWN() from <util.h>, which does not work with ASLR.
+ */
+#define LD_ROUNDUP(x, y)  (x) + (((x) % (y)) ? ((y) - ((x) % (y))) : 0)
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S:68]: Out of curiosity, did you attempt to implement these two based on the builtin function `ALIGN()`?

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/kern.ld.S (hunk 1 above)
Comment (jforissier) [reply core/arch/arm/kernel/kern.ld.S]: No I didn't, good suggestion. Updated.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

File: core/arch/arm/kernel/kern.ld.S [hunk 2]
```diff
@@ -61,6 +61,13 @@
 #define TEE_TEXT_VA_START       (TEE_RAM_VA_START + \
 					(TEE_LOAD_ADDR - TEE_RAM_START))
 
+/*
+ * Note:
+ * Clang 11 (ld.lld) generates non-relocatable reference when using ROUNDDOWN()
+ * from <util.h>, which does not work with ASLR.
+ */
+#define LD_ROUNDDOWN(x, y) (x) - ((x) % (y))
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/kern.ld.S]: Right, it is good practice. Added.

**Comment by jforissier** [comment]
`ALIGN()` used, squashed and slightly reworded the commit description. @jenswi-linaro I have kept your R-b, please shout if you think something is wrong ;-)

File: core/arch/arm/kernel/kern.ld.S [hunk 3]
```diff
@@ -61,6 +61,13 @@
 #define TEE_TEXT_VA_START       (TEE_RAM_VA_START + \
 					(TEE_LOAD_ADDR - TEE_RAM_START))
 
+/*
+ * Note:
+ * Clang 11 (ld.lld) generates non-relocatable reference when using ROUNDDOWN()
+ * from <util.h>, which does not work with ASLR.
+ */
+#define LD_ROUNDDOWN(x, y) ((x) - ((x) % (y)))
```
Comment (etienne-lms) [inline core/arch/arm/kernel/kern.ld.S:69]: maybe a tabulation before the macro value

Comment (jforissier) [reply core/arch/arm/kernel/kern.ld.S:69]: A tab would not improve anything here, there is nothing to align with.

**Pull request merged**